many: Vendor apparmor-3.0.3 into the snapd snap (take 2)

alexmurray opened the pull request:
This PR resubmits #9939 since it subsequently got reverted in #11090.

This was reverted as `snapd-apparmor` failed to use the vendored `apparmor_parser` and so would fail to properly load AppArmor policies as it would always use the system installed `apparmor_parser` which may not be compatible with the vendored one that was used to generate the policies in the first place.

To fix this, I have reimplemented `snapd-apparmor` in Go so it can compile policy in the same way as snapd itself and hence will use the correct vendored AppArmor when it is present.

mardy commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,149 @@
+)
+
+// Checks to see if the current container is capable of having internal AppArmor
+// profiles that should be loaded. Callers of this function should have already
+// verified that they're running inside of a container environment with
+// something like `systemd-detect-virt --container`.
```
This does not seem to be the case, since this function is executing this check itself.

alexmurray replied:
Ah thanks, that's what I get for copy-pasta-ing the comment from the bash version into the Go impl.

mardy commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,149 @@
+		if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
+			return false
+		}
+
```
nitpick: an empty killer line! ;-)

alexmurray replied:
Thanks

mardy commented on interfaces/apparmor/apparmor.go:
```diff
@@ -117,10 +118,21 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
+		// - same for if we are not using the internal
+		// apparmor_parser so that we can support a mocked parser
+		// during tests
+		parser = "apparmor_parser"
```
should we emit a `log.Noticef("Using apparmor_parser from $PATH")` or similar, if this can help debugging?

alexmurray replied:
I don't think we need to - as part of the apparmor vendoring work I added a new sandbox-feature `parser:snapd-internal` which is then present in the `snap debug sandbox-features` output when using the vendored apparmor - so we can check for this when debugging instead.

mardy commented on sandbox/apparmor/apparmor.go:
```diff
@@ -316,43 +321,92 @@ func probeKernelFeatures() ([]string, error) {
+		if osutil.IsExecutable(path) {
+			prefix := strings.TrimSuffix(path, "apparmor_parser")
+			// when using the internal apparmor_parser also use
+			// it's own configuration and includes etc plus
```
super nitpick: its

alexmurray replied:
Good point but this is in the original commit which I would rather restore with a single "Revert" commit so I don't think it's worth fixing at this point with an extra commit.

mardy commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,149 @@
+// In addition the script assumes that the system-wide apparmor service has
+// already executed, initializing apparmor file-systems as necessary.
+//
+// NOTE: This script ignores failures in some scenarios as the intent is
+// to simply load application profiles, as many as we can, even if for
+// whatever reason some of those fail.
```
It's stil not clear to me why failures are not treated as errors. Is it because this script is just a mean to speed-up the execution of snaps by pre-loading their profile? If so, I'd add a "for performance purposes" somewhere in this sentence.

alexmurray replied:
Thanks will do.

mardy reviewed (approved):
LGTM, just minor comments and questions :-)

pedronis commented:
does the last changes fully fix the problem? do we need to implement reexec in snapd-apparmor? are we sure the snapd snap is mounted when snapd-appamor running?

my worry is that snapd-apparmor will come from the deb on classic, and the code we added in apparmor depends on searching tools relative to /proc/self/exe so I don't know if in the end it will find the snapd apparmor_parser ?

alexmurray commented:
> does the last changes fully fix the problem? do we need to implement reexec in snapd-apparmor? are we sure the snapd snap is mounted when snapd-appamor running?

Thanks - added reexec in https://github.com/snapcore/snapd/pull/11096/commits/dfd7875168a2cb01f361f37d69e1584c4a3387a2

> 
> my worry is that snapd-apparmor will come from the deb on classic, and the code we added in apparmor depends on searching tools relative to /proc/self/exe so I don't know if in the end it will find the snapd apparmor_parser ?

This is a great point - thanks, I hadn't considered this :) Hopefully the reexec deals with this and I have also extended the spread test to be more comprehensive (plus the snapd-snap spread test also restarts snapd.apparmor.service which should help to test the re-exec case as well).

pedronis commented:
@alexmurray @anonymouse64 

> does the last changes fully fix the problem? do we need to implement reexec in snapd-apparmor? are we sure the snapd snap is mounted when snapd-appamor running?
>
> Thanks - added reexec in dfd7875

thanks, there's another twist I fear. As the snapd snap and the classic packaging are not released fully in sync everywhere, we cannot use on classic the apparmor_parser from the snapd snap unless we can check that the host system has a new reexecing snapd-apparmor

anonymouse64 commented:
So I thought a bit about this PR and what Samuele mentioned above and I can think of two ways to handle this:

1. We could change the service unit for snapd.apparmor.service to specify something like X-Snappy=foo or whatever and then snapd when deciding to use the apparmor_parser from snapd or not could check for that to decide (if that is absent, then the snapd.apparmor.service is the old one, and thus will fail the way we saw it before, so in that case we would keep using apparmor_parser from the host instead of the bundled one).
2. We could add something to snapd-apparmor itself wherein that binary/script is called to ask it if it supports re-exec. The current shell script has the nice advantage that it understands a single command "start", so if we call it with `snapd-apparmor re-exec-capable`, the current script will do nothing, whereas the new Go implementation of snapd-apparmor could output something like "true", etc to indicate that it does support re-exec and thus snapd is okay to use the apparmor_parser from the snapd snap instead of the one from the host.

I think I prefer option 2 personally, but curious what others think about this proposal.

Also, @alexmurray how do you feel about splitting out the snapd-apparmor changes to a separate PR since they don't _necessarily_ need all the apparmor stuff in the snapd snap and it would be easier to review those changes separate from the original revert of the revert here? Not all of the features of snapd-apparmor reimplementation would be testable in a separate PR, so some of those tests would have to still be reviewed in this PR, but I just fear that this PR has again the same problem as before if it grows in complexity...

anonymouse64 commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,157 @@
+// In addition the script assumes that the system-wide apparmor service has
+// already executed, initializing apparmor file-systems as necessary.
+//
+// NOTE: This script ignores failures in some scenarios as the intent is to
```
not really a script anymore :-)

anonymouse64 commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,157 @@
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(2)
+	}
+	os.Exit(0)
```
nitpick, but this is unnecessary

anonymouse64 commented on cmd/snapd-apparmor/main_test.go:
```diff
@@ -0,0 +1,122 @@
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	appArmorSecurityFSPath := filepath.Join(dirs.GlobalRootDir, "/sys/kernel/security/apparmor/")
+	if err := os.MkdirAll(appArmorSecurityFSPath, 0755); err != nil {
+		panic(err)
+	}
```
```suggestion
	err := os.MkdirAll(appArmorSecurityFSPath, 0755)
	c.Assert(err, IsNil)
```

anonymouse64 commented on cmd/snapd-apparmor/main_test.go:
```diff
@@ -0,0 +1,122 @@
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	f, err := os.Create(filepath.Join(appArmorSecurityFSPath, ".ns_stacked"))
+	if err != nil {
```
same

anonymouse64 commented on cmd/snapd-apparmor/main_test.go:
```diff
@@ -0,0 +1,122 @@
+	c.Assert(err, IsNil)
+
+	// catch unexpected changes to apparmor_parser compiler flags etc
+	testutil.MockCommand(c, "apparmor_parser", `echo "$@"; ""exit "$#"`)
```
this can be done in a simpler way where we want to check the arguments that are passed:
```suggestion
	parserCmd := testutil.MockCommand(c, "apparmor_parser", `exit 1`)
	defer parserCmd.Restore()
	
	...
	
	c.Assert(parserCmd.Calls(), DeepEquals, [][]string{
		{"apparmor_parser", "foo-args"},
	})
```

this can also be used to check that it is successful if you use `exit 0` or just `""` as the script in MockCommand too

anonymouse64 commented on cmd/snapd-apparmor/main_test.go:
```diff
@@ -0,0 +1,122 @@
+	}
+	err = snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+}
```
we also usually like to have a test which actually executes main(), see some of the examples in cmd/snap-repair/*test.go for an example, but one thing that sort of test would catch which we don't currently have coverage for here is the systemd-detect-virt invocation for example

anonymouse64 replied:
doing this may require using go-flags, but I think that's a completely fair thing to do

anonymouse64 commented on cmd/snapd-apparmor/main.go:
```diff
@@ -0,0 +1,157 @@
+	"github.com/snapcore/snapd/snapdtool"
+)
+
+// Checks to see if the current container is capable of having internal AppArmor
```
should we have a comment somewhere like we had in the original shell script that this code comes from apparmor stuff in ubuntu? (though here obviously it is not copied it is adapted)

anonymouse64 commented on tests/main/snapd-apparmor/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    journalctl -u snap.apparmor.service
+
+execute: |
+    if ! systemctl is-active snapd.apparmor.service; then
```
where is the service not active? if we don't want to run this test on i.e. systems without apparmor at all, perhaps we should instead make this a check that on systems like fedora etc that snapd.apparmor.service is not present at all?

anonymouse64 replied:
also we have struggled to identify what distros we expect to see snapd.apparmor.service on, and it seems we now essentially expect to find it everywhere we have apparmor, so it would be nice if this test could also essentially serve as documentation for that in that it has a comment and associated checks that we have snapd.apparmor.service on distros that should have it. Currently you have to go digging through the packaging rules to identify whether the service is included or not in a given distro

anonymouse64 commented on tests/main/snapd-apparmor/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    "$TESTSTOOLS"/snaps-state install-local "$CONSUMER_SNAP"
+    tests.cleanup defer snap remove --purge "$CONSUMER_SNAP"
+
+    # stop snapd so it does not try and load apparmor policies directly
```
snapd will already have loaded policies as part of installing the snap, maybe

```suggestion
    # stop snapd so we can unload the apparmor policies manually
```

anonymouse64 commented on tests/main/snapd-apparmor/task.yaml:
```diff
@@ -0,0 +1,73 @@
+
+    # also check that snapd.apparmor.service fails when a profile is invalid
+    sed -i s/profile/profileinvalidnametobereplaced/ /var/lib/snapd/apparmor/profiles/snap.$CONSUMER_SNAP.*
+    tests.cleanup defer sed -i s/profileinvalidnametobereplaced/profile/ /var/lib/snapd/apparmor/profiles/snap.$CONSUMER_SNAP.*
```
thanks for this test too!

anonymouse64 reviewed (commented):
overall is looking good, but please see my other comment about proposals on how to handle the classic packaging vs snap mismatch issue that @pedronis raised and also possibly splitting the snapd-apparmor re-write into another PR

thanks for the spread test too!

alexmurray commented:
@anonymouse64 I agree splitting that it makes sense to split out the snapd-apparmor reimpl from this PR so I'll take a stab at that against current master.

Regarding the other suggestions - I am not sure I understand what you mean - snapd-apparmor is independent from snapd (we are just reusing a bunch of the code from snapd now) - and snapd doesn't call into snapd-apparmor, so I am not sure how having it describe whether it is capable of re-exec would be useful. But I wouldn't be surprised if I am missing something here so please help me understand your suggestion.

Personally, I feel like the best solution here would be to do snapd-apparmor reimpl via a separate PR as you suggest *and* include the re-exec support in that. This should then get merged to master *and* a release branch so that we can do a SRU release of snapd to Ubuntu which then includes the re-exec support for snapd-apparmor. Thus when the vendored apparmor stuff eventually does land in the snapd snap, we can support re-exec into it from the snapd deb and with any luck it should all just work.

So I'll split this up and try addressing your comments for snapd-apparmor in a separate PR and once that makes progress we can come back to this one.

alexmurray commented:
Ok see #11129 for now - this splits the snapd-apparmor bits out from this PR - let's focus on that first then we can come back to this.

anonymouse64 commented:
@alexmurray the issue that Samuele explained is that snapd-apparmor on the host will come from the traditional linux package for snapd (on classic at least), so even if we update snapd-apparmor inside the snapd snap, that version of snapd-apparmor is not installed on the host when we install the new snapd snap, so the service file for snapd.apparmor.service will not use our new snapd-apparmor that we ship from the snapd snap. 

This leads to the bug we had after we merged your initial PR, where now snapd inside the snapd snap (since we do properly re-exec the snapd daemon from the deb world to the snapd snap world) thinks it can use the parser from the snapd snap since the parser is now available from the snapd snap, but this is not enough. The snapd snap needs to consider snapd-apparmor on the host from the traditional linux package to know if it can write policy using the parser shipped inside the snapd snap, very specifically we need to implement this logic:
```go
func shouldUseSnapdInternalAppArmorParser() bool {
    if snapdSnapHasInternalParser() {
        if snapdHostAppArmorServiceSupportsReExec() {
             return true
        }
    }
    return false
}
```

because otherwise even if snapd-apparmor inside the snapd snap supports re-exec, snapd.apparmor.service will still be broken until both the snapd snap AND the snapd deb/rpm/etc is updated with the new snapd-apparmor.

The other alternative to this complexity that I don't like is that we could implement re-exec inside snapd-apparmor without allowing using the parser from the snapd snap and then just wait a long time to be "pretty sure" that all the traditional linux packages of snapd have been updated and then turn on using the parser from inside the snapd snap, but this still has the problem that even if the new traditional linux package is available, we can't force folks to be up to date with it the same way we can with the snapd snap via the store.

Hopefully this all makes sense.

pedronis commented:
@anonymouse64 @alexmurray 

> So I thought a bit about this PR and what Samuele mentioned above and I can think of two ways to handle this:

there's a 3rd option, add a flag to data/info => $libexecdir/snapd/info (created by mkversion.sh)  like SNAPD_APPARMOR_GO=1 or similar and then add a helper to snapdtool to read this kind of flags parallel to SnapdVersionFromInfoFile.  We should probably teach these helpers to take a dir instead of the full path though.

anonymouse64 commented:
@pedronis I implemented something like the third option you mention in https://github.com/snapcore/snapd/pull/11137

alexmurray commented:
Closing in favor of #12081 12087
